java-checks/src/main/java/org/sonar/java/checks/ThreadLocalCleanupCheck.java

```diff
@@ -0,0 +1,89 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import java.util.Arrays;
+import java.util.List;
+import org.sonar.check.Rule;
+import org.sonar.java.matcher.MethodMatcher;
+import org.sonar.java.matcher.TypeCriteria;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+@Rule(key = "S5164")
+public class ThreadLocalCleanupCheck extends IssuableSubscriptionVisitor {
+
+  private static final String THREAD_LOCAL = "java.lang.ThreadLocal";
+  private static final MethodMatcher THREADLOCAL_SET = MethodMatcher.create()
+    .typeDefinition(THREAD_LOCAL).name("set").addParameter(TypeCriteria.anyType());
+  private static final MethodMatcher THREADLOCAL_REMOVE = MethodMatcher.create()
+    .typeDefinition(THREAD_LOCAL).name("remove").withoutParameter();
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Arrays.asList(Tree.Kind.CLASS, Tree.Kind.METHOD_INVOCATION);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+    if (tree.is(Tree.Kind.CLASS)) {
+      Symbol.TypeSymbol clazz = ((ClassTree) tree).symbol();
+      clazz.memberSymbols().stream()
+        .filter(Symbol::isVariableSymbol)
+        .filter(s -> s.type().is(THREAD_LOCAL))
+        .filter(ThreadLocalCleanupCheck::isNotSuper)
+        .forEach(this::checkThreadLocalField);
+    } else {
+      MethodInvocationTree mit = (MethodInvocationTree) tree;
+      if (THREADLOCAL_SET.matches(mit) && mit.arguments().get(0).is(Tree.Kind.NULL_LITERAL)) {
+        reportIssue(mit, "Use \"remove()\" instead of \"set(null)\".");
+      }
+    }
+  }
+
+  private static boolean isNotSuper(Symbol s) {
+    return !"super".equals(s.name());
+  }
+
+  private void checkThreadLocalField(Symbol field) {
+    if (field.usages().stream().noneMatch(ThreadLocalCleanupCheck::usageIsRemove)) {
+      reportIssue(((VariableTree) field.declaration()).simpleName(), "Call the \"remove()\" on \"" + field.name() + "\" in a \"finally\" block.");
+    }
+  }
+
+  private static boolean usageIsRemove(IdentifierTree usage) {
+    Tree parent = usage.parent();
+    if (parent.is(Tree.Kind.MEMBER_SELECT)) {
+      Tree mseParent = parent.parent();
+      return mseParent != null && mseParent.is(Tree.Kind.METHOD_INVOCATION)
```

Comment:
not sure about the null check here (I know that you are getting it because of the annotation on the `parent()` method), but it will never be possible here. Maybe `Objects.requireNonNull(parent.parent()).is(Tree.Kind.METHOD_INVOCATION)`?

Reply:
tbh, I find null check more readable that the alternative you are suggesting. I prefer to be bit more defensive than to risk NPE if API contract changes